types/transactions.go

```diff
@@ -143,17 +144,51 @@ type (
 // ID returns the id of a transaction, which is taken by marshalling all of the
 // fields except for the signatures and taking the hash of the result.
 func (t Transaction) ID() TransactionID {
-	return TransactionID(crypto.HashAll(
-		t.SiacoinInputs,
-		t.SiacoinOutputs,
-		t.FileContracts,
-		t.FileContractRevisions,
-		t.StorageProofs,
-		t.SiafundInputs,
-		t.SiafundOutputs,
-		t.MinerFees,
-		t.ArbitraryData,
-	))
+	var hash crypto.Hash
+	h := crypto.NewHash()
+	enc := encoding.NewEncoder(h)
+
+	for i := range t.SiacoinInputs {
+		t.SiacoinInputs[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.SiacoinOutputs)))
+	for i := range t.SiacoinOutputs {
+		t.SiacoinOutputs[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.FileContracts)))
+	for i := range t.FileContracts {
+		enc.Encode(t.FileContracts[i])
+	}
+	encoding.WriteInt(h, len((t.FileContractRevisions)))
+	for i := range t.FileContractRevisions {
+		enc.Encode(t.FileContractRevisions[i])
+	}
+	encoding.WriteInt(h, len((t.StorageProofs)))
+	for i := range t.StorageProofs {
+		enc.Encode(t.StorageProofs[i])
+	}
+	encoding.WriteInt(h, len((t.SiafundInputs)))
+	for i := range t.SiafundInputs {
+		enc.Encode(t.SiafundInputs[i])
+	}
+	encoding.WriteInt(h, len((t.SiafundOutputs)))
+	for i := range t.SiafundOutputs {
+		t.SiafundOutputs[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.MinerFees)))
+	for i := range t.MinerFees {
+		t.MinerFees[i].MarshalSia(h)
+	}
+	encoding.WriteInt(h, len((t.ArbitraryData)))
+	for i := range t.ArbitraryData {
+		encoding.WritePrefix(h, t.ArbitraryData[i])
+	}
+	encoding.WriteInt(h, len((t.TransactionSignatures)))
+	for i := range t.TransactionSignatures {
```

Comment:
The biggest reason is because when you do

`for i := range myArray`, you get to access each element in `myArray` directly, so there's no copying or anything. When you do

`for _, obj := range myArray`, it allocates new space for `obj` and copies each element of the array into `obj` when you access it inside the loop. So if you modify obj, it won't modify the actual memory in `myArray` (note: if `obj` contains pointers, then you can still change values that will also affect the values seen by `myArray`).

So the primary reason that it's faster (from a high level, anyway) is because you get to skip on 1 allocation of size `obj` and you get to skip on `len(myArray)` copy calls of size `obj`. It doesn't create much GC pressure but it does create a lot of copying issues.

That's also why you can't do this:

```
for _, obj := range myArray {
    go manipulateObj(&obj)
}
```

You'll get a bunch of race conditions because each goroutine is pointing at the same spot in memory.